Diff View: Match similar lines in changed blocks

mstv opened the pull request:
## Proposed changes

- Detect indented lines
- Then match by common words
- Reduce the highlight (from git) for identical parts of the line - instead of marking the different part
- Remove highlighting of filenames in the diff header because their background highlighting was removed in f201220e8b77f5ab218277f60fd3e008c1ee5041
- Some refactoring into functions

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

`FormRevertCommit.cs` in 6a9f3c01b1c988d5a01369a5f54020115022a5cf

![image](https://github.com/user-attachments/assets/d0ff86e4-ffce-41f8-ad03-10c954e7215d)

### After (final variant)

![grafik](https://github.com/user-attachments/assets/2e096864-b8cd-4657-81a0-f67319471e89)

### After (intermediate variant with bold highlight)

![image](https://github.com/user-attachments/assets/678630e4-6053-4040-b4db-296ab8dc1946)

## Test methodology <!-- How did you ensure quality? -->

- some additional tests

## Please do not squash merge

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
Why not mark the new lines like 97,98,108 ?

mstv commented:
> Why not mark the new lines like 97,98,108 ?

line 111, too
I thought about it. But I did not want to change that part of the existing behavior (refer to lines 103-104, 110-111. which have always been the last lines - being ignored because more lines were added than removed).
It will be clearer to mark the entire unmatched lines - prepared in branch `feat/multiline_diff-2` 698acf6cde1a2e7e5c1717f4689727ddddbcdad8.
Looks good for the example chosen for the PR description:

![image](https://github.com/user-attachments/assets/68360c06-bd88-4e74-8efd-a0c3154864c3)

Though I am not fully convinced because of the different display of really added blocks:

![image](https://github.com/user-attachments/assets/30f08a5e-eea9-43d1-8a35-ea016a0bd8e9)

![image](https://github.com/user-attachments/assets/ed5d45ae-ec64-4c74-8453-34b2ae7c023f)

gerhardol commented:
What about using a dimmed foreground for normal?
An example below, all added blocks do not get non dimmed foreground
As the color is selected by Git, the color has to be set directly (hardcoded for now, could be derived from the normal GE color detection)

![image](https://github.com/user-attachments/assets/475a0f04-641e-49ad-8cd7-22bc591ccb7e)

Patch
![image](https://github.com/user-attachments/assets/ffdb6653-46b7-4394-a2d5-3ff97723393e)

gerhardol commented:
It is probably only the unchanged parts that should be dimmed. The other lines are added in some sense, my patch handles lines within blocks where similarities were found and normal and completely new blocks as dim (like -54-56 , + 56-60 in the first example. 
So the PR as is is better than feat/multiline_diff-2 or my patch. 

But dimming the non changed parts where there are changes (all other diffs in the example is an improvement for me, it is easier to ignore existing code. That would be implemented with dimming the remaining line not set to bold, based on this PR. (GetDifferenceMarkers marks the complete line as dimmed after setting the changed part would work.)

mstv commented:
Thank you for nudging in the direction of less highlighting!
It will make perfect sense to _remove_ the color from identical parts - instead of adding another kind of highlight. It will look more like git word diff.
I also have an idea how to mark the place of pure insertions/removals.

mstv commented:
> _remove_ the color from identical parts - instead of adding another kind of highlight.

I have updated the PR description.

> I also have an idea how to mark the place of pure insertions/removals.

It could look like in the following screenshots. But I would like to discuss and implement this in a follow-up.

![image](https://github.com/user-attachments/assets/4ed18188-340d-4b0e-9c9e-f726f7719e4d)

![image](https://github.com/user-attachments/assets/421f1385-6a93-4e01-9e25-cf8ef2514f86)

mstv commented:
> The algorithm tries to color the header too. There are no coloring of the header right now.

"Right now" was a good one. I scratched my head how I could have broken it - I haven't but f201220e8b77f5ab218277f60fd3e008c1ee5041 of #11590.

gerhardol commented:
@gitextensions/git-extensions-team Are you OK with the change of presentation for diff, where the pre/post parts of lines where no change has been detected at all is not marked at all? 
There is no more bold markers, changed parts of a line are presented as completely new/removed lines.

I may want to have the unchanged parts with dimmed background or foreground (text) to see the changed lines. It may change when I learn to navigate though.

Will review when I have decided what I would like to see.
This is already a great improvement.

mstv commented:
(a) no mark | (b) pale
--|--
![image](https://github.com/user-attachments/assets/93e721c5-284a-4544-8543-66d7d6110c6c) | ![grafik](https://github.com/user-attachments/assets/2e096864-b8cd-4657-81a0-f67319471e89)

gerhardol commented:
> (b) pale

What I prefer, at least as a transition

RussKie commented:
I prefer (b) too.

gerhardol commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -311,32 +311,35 @@ private static IEnumerable<TextMarker> GetDifferenceMarkers(Func<int, char> getC
-        static TextMarker CreateUnmarker(int offset, int length)
-            => CreateTextMarker(offset, length, AppColor.AnsiTerminalWhiteBackBold.GetThemeColor());
+        static TextMarker CreatePaleMarker(int offset, int length, Color color)
+            => CreateTextMarker(offset, length, new HslColor(color).WithLuminosity(0.5 + (0.47 * (ColorHelper.IsLightTheme() ? 1 : -1))).ToColor());
```
Add the color transformer to ColorHelper?

This is slightly brighter than the DimColor that can be set in the terminal. Should these be the same?
The Dim implementation is in AnsiUtilities right now, could be moved to ColorHelper.
Great change!

mstv replied:
> This is slightly brighter than the DimColor that can be set in the terminal.

I am going to try that. But I did not like it with lower luminosity.

mstv replied:
I have refactored it with `DimColor` and have applied it twice.

gerhardol commented:
The matching algoritm is much better, but not perfect.
The diff looks strange for 226/227

![image](https://github.com/user-attachments/assets/0a50878a-5815-4c1c-9ff5-aa1d14735bbe)

TortoiseDiff handles that better

![image](https://github.com/user-attachments/assets/4b45ef53-7040-4b43-9cc7-8b7720b87b51)

master is better in this situation (below is .17886, so the extra marker is there)

![image](https://github.com/user-attachments/assets/62bb7533-0db3-4eb7-b2ae-033561fb6571)

gerhardol reviewed (commented):
+2
Not any comments on the code, partly due to that I do not understand the intention of the algorithm. But I do not know how to document it, describing each method is probably not helping. I guess I will not have a formal request and approve formally within the next days.

mstv commented:
I have rebased in order to resolve the MC, squashed the fixups and added 2.5 comments. (For comparison, I have kept the previous state in branch `mstv/feat/inline_diff-before_rebase`.)

gerhardol commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -272,26 +284,66 @@ private static void MarkDifference(IDocument document, ISegment lineRemoved,
+        static Color GetRemovedBackColor() => AppColor.AnsiTerminalRedBackNormal.GetThemeColor();
     }
 
     private void AddExtraPatchHighlighting(IDocument document)
```
Maybe rename this function
On 320/372 add comment that the file names in the header is updated
On 323/375 describe the function briefly, similar to the PR description. Maybe in GetDifferenceMarkers instead.
Then you have a starter to understand and improve the algorithm even further

gerhardol reviewed (approved):
great

gerhardol commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -317,18 +370,18 @@ private void AddExtraPatchHighlighting(IDocument document)
             }
 
-            MarkDifference(document, linesRemoved, linesAdded, diffContentOffset);
+            MarkInlineDifferences(document, linesRemoved, linesAdded, diffContentOffset);
```
I prefer to not mark the filenames in the headers.
Removed in my branch feature/difflineinfo-coloring. (The filename in the header is also not parsed now, will need to be added.)

mstv replied:
Removed

gerhardol reviewed (commented):
I want this included in 5.0 or a patch release

RussKie commented:
> I want this included in 5.0 or a patch release

👍 There are few items pending.

gerhardol commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -272,63 +280,89 @@ private static void MarkDifference(IDocument document, ISegment lineRemoved,
             found = false;
-            linesRemoved = GetRemovedLines(document, ref line, ref found);
-            linesAdded = GetAddedLines(document, ref line, ref found);
+            List<ISegment> linesRemoved = GetRemovedLines(document, ref line, ref found);
```
With the current implementation, there is no reason to run MarkInlineDifferences if not found (and not GetAddedLines() if not found in GetRemovedLines(), that is no mayor thing if only-parse once in my follow up is used).

But the algorithm could be expanded, if there are only few lines between two "removed" blocks (maybe just newlines) and similar for added, the blocks could be used for the context.

This could well be done in a follow-up too.

Git detects "if(EXISTS ${download_directory})" as moved.
There is no dimming on the correct line for "file(LOCK ${download_directory}/../cmake.lock RELEASE)" as moved with indentation is in the second block (dimming on file(READ ${download_directory}.download expected_checksum))

```patch
-
-    if(EXISTS ${download_directory})
-      if(CPM_SOURCE_CACHE)
-        file(LOCK ${download_directory}/../cmake.lock RELEASE)
+    if(EXISTS ${download_directory}
+      AND CPM_ARGS_CUSTOM_CACHE_CHECKSUM_COMMAND
+      AND (CPM_CHECK_CACHE_CHECKSUM
+        OR DEFINED CPM_ARGS_CUSTOM_CACHE_CHECKSUM_VALUE))
+      if (CPM_ARGS_CUSTOM_CACHE_CHECKSUM_VALUE)
+        # Explicit checksum provided, ignore value in .downloaded
+        set(expected_checksum ${CPM_ARGS_CUSTOM_CACHE_CHECKSUM_VALUE})
+      else()
+        file(READ ${download_directory}.download expected_checksum)
+        string(STRIP "${expected_checksum}" expected_checksum)
       endif()
 
+    if(EXISTS ${download_directory})
+      # Directory content is considered OK
+      file(LOCK ${download_directory}/../cmake.lock RELEASE)
       cpm_store_fetch_properties(
 ```

mstv replied:
I have pushed the evaluation of `found`.
Matching non-contiguous blocks / enlarging blocks is orthogonal to this PR, i.e. should be done in a follow-up.

gerhardol replied:
Yes, this was just a discussion
I pushed an experiment with allowing three lines of gap in c1ac2643afebe266143494c1b988701a9066b63c. That could probably be at least 5, it is only with "View complete file" that the parse does not hit a @@ Header.

mstv commented:
This PR is ready and is blocking two dependent PRs.
Gerhard and me ask for a second review.

RussKie reviewed (commented):
It would be great to have a broader test coverage.

gerhardol commented on src/app/GitExtUtils/GitUI/Theming/ColorHelper.cs:
```diff
@@ -13,6 +13,17 @@ private static readonly (KnownColor back, KnownColor fore)[] BackForeExamples =
+        /// <summary>
+        ///  Blends the color with the default background color, halves each value first.
+        /// </summary>
+        public static Color DimColor(Color color)
```
Dim colors is a terminal color concept

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -272,63 +280,102 @@ private static void MarkDifference(IDocument document, ISegment lineRemoved,
+        static TextMarker CreateTextMarker(int offset, int length, Color color)
+            => new(offset, length, TextMarkerType.SolidBlock, color, ColorHelper.GetForeColorForBackColor(color));
+
+        static Color GetAddedBackColor() => AppColor.AnsiTerminalGreenBackNormal.GetThemeColor();
+        static Color GetRemovedBackColor() => AppColor.AnsiTerminalRedBackNormal.GetThemeColor();
```
There will be more colors.
Do you really want empty lines in-between?
```
        static Color GetAddedBackColor() => AppColor.AnsiTerminalGreenBackNormal.GetThemeColor();
        static Color GetAddedForeColor() => AppColor.AnsiTerminalGreenForeBold.GetThemeColor();
        static Color GetRemovedBackColor() => AppColor.AnsiTerminalRedBackNormal.GetThemeColor();
        static Color GetRemovedForeColor() => AppColor.AnsiTerminalRedForeBold.GetThemeColor();
```

mstv replied:
`GetAddedBackColor()` etc. turned into `private static` fields.

I am going to squash the fixup commits and merge it.

pmiossec reviewed (approved):
Briefly reviewed.
I have not seen things standing out.

Result seems great 👍. Especially on one-line changes but not always used to on big multi-lines changes....

gerhardol commented:
> It would be great to have a broader test coverage.

Can we wait with that?

mstv #11862 depends in this, I have some pending changes after this that refactors this setup (and removes parsing).
I had planned to test part of current MarkInlineDifferences().

mstv commented:
I have added some function tests, (which should remain valid - at least as a basis).

gerhardol commented on tests/app/UnitTests/GitUI.Tests/Editor/Diff/DiffHighlightServiceTests.cs:
```diff
@@ -0,0 +1,50 @@
+        TextMarker[] expectedMarkers =
+        [
+            new TextMarker(offset: addedLine.Offset + lineStartOffset, length: identicalPartBefore.Length, TextMarkerType.SolidBlock),
+            new TextMarker(offset: addedLine.Offset + lineStartOffset + identicalPartBefore.Length + differentAdded.Length, length: identicalPartAfter.Length, TextMarkerType.SolidBlock),
```
```suggestion
            new TextMarker(offset: addedLine.Offset + lineStartOffset + identicalPartBefore.Length + differentAdded.Length, length: identicalPartAfter.Length, dontCare),
```

mstv replied:
`dontCare` does not apply here.
The `MarkerType` of the results should be verified. The improved version in the next PR will not work without it.

gerhardol commented on tests/app/UnitTests/GitUI.Tests/Editor/Diff/DiffHighlightServiceTests.cs:
```diff
@@ -0,0 +1,50 @@
+        TextMarker addedLine = new(offset: text.IndexOf(addedLineText), addedLineText.Length, textMarkerType: dontCare);
+        const int lineStartOffset = 1;
+
+        IEnumerable<TextMarker> markers = DiffHighlightService.GetDifferenceMarkers(GetCharAt, removedLine, addedLine, lineStartOffset);
```
I had planned to test the refactored GetBlockOfLines() in my branch, the test for "merged gaps" can be done more easily.
That will make more extensive tests possible as a compliment.
(Combined use that use not adjacent blocks will require tests like this too.)

gerhardol commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -13,6 +13,9 @@ namespace GitUI.Editor.Diff;
 /// </summary>
 public abstract class DiffHighlightService : TextHighlightService
 {
+    private static readonly Color _addedBackColor = AppColor.AnsiTerminalGreenBackNormal.GetThemeColor();
```
The theme color is basically a dictionary lookup if I remember correctly. Is this needed?

mstv replied:
These fields are used more than two times each. Thus the "rule of three" applies.
Theme changes require a restart of GE anyway.
